### PR TITLE
Set up useBackHandler hook

### PR DESCRIPTION
**Pull request by asgalecki**

### CODE REVIEW

**Comment by asgalecki** [comment]
https://user-images.githubusercontent.com/70109479/173818255-50d59cf8-e970-4113-9b72-602b77243ed4.mp4

File: src/components/RequestDetails/SeeRequest.tsx [hunk 1]
```diff
@@ -27,6 +28,14 @@ export const SeeRequest = ({ route: { params: p } }: RequestsNavigationProps<'SE
     Analytics().track('REQUEST_OPENED', { request: { ...p } })
   }, [p])
 
+  useBackHandler(() => {
+    if (prevScreen) {
+      navigation.navigate(prevScreen)
+      return true
+    }
+    return false
+  })
```
Comment (iceu-bb) [inline src/components/RequestDetails/SeeRequest.tsx]: This logic is repeated in several places. Please move it to custom hook e.g `usePrevScreenBackHandler`

File: src/screens/calendar/Calendar.tsx [hunk 2]
```diff
@@ -11,11 +11,17 @@ import { getISODateString, parseISO } from 'utils/dates'
 import { RequestsContextProvider } from 'contexts/RequestsProvider'
 import { useBooleanState } from 'hooks/useBooleanState'
 import { LoadingModal } from 'components/LoadingModal'
+import { useBackHandler } from 'hooks/useBackHandler'
+import { RouteProp, useFocusEffect, useNavigation, useRoute } from '@react-navigation/native'
+import { BottomTabRoutes } from 'navigation/types'
 import { CategoriesSlider } from './components/CategoriesSlider'
 
 const CalendarToWrap = () => {
+  const [prevScreen, setPrevScreen] = useState('')
```
Comment (iceu-bb) [inline src/screens/calendar/Calendar.tsx]: `prevScreen` shouldn't be part of component state, it's navigation responsibility.

File: src/screens/calendar/Calendar.tsx [hunk 3]
```diff
@@ -11,11 +11,17 @@ import { getISODateString, parseISO } from 'utils/dates'
...
   const [currentIndex, setCurrentIndex] = useState(0)
   const flatListRef = useRef<FlatList>(null)
+  const route = useRoute<RouteProp<BottomTabRoutes, 'CALENDAR'>>()
+  const navigation = useNavigation()
```
Comment (iceu-bb) [inline src/screens/calendar/Calendar.tsx]: missing type for `useNavigation` hook

**Review by iceu-bb — CHANGES_REQUESTED** [review]

File: src/screens/calendar/Calendar.tsx (hunk 2 above)
Comment (asgalecki) [reply src/screens/calendar/Calendar.tsx]: I need this useState to hold the prevState value as I need to clear the params.prevState and make it undefined. Without it the back handler doesn't work properly:

https://user-images.githubusercontent.com/70109479/173858555-9e2ce324-ed6d-41ec-9528-31cfea25974a.mp4

Comment (asgalecki) [reply src/screens/calendar/Calendar.tsx]: And version with useState used to cache params.prevScreen:

https://user-images.githubusercontent.com/70109479/173858771-7ef2ea63-4ddb-4e63-b75c-4a4d4940c4d2.mp4

File: src/hooks/usePrevScreenBackHandler.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import { NavigationProp, ParamListBase } from '@react-navigation/native'
+import { useBackHandler } from './useBackHandler'
+
+type PrevScreens = 'NOTIFICATIONS' | 'STATS_AND_REQUESTS'
+
+export const usePrevScreenBackHandler = (
+  navigation: NavigationProp<ParamListBase>,
```
Comment (iceu-bb) [inline src/hooks/usePrevScreenBackHandler.ts]: Can you take it from `useNavigation` hook ?

File: src/screens/calendar/Calendar.tsx [hunk 5]
```diff
@@ -11,11 +11,19 @@ import { getISODateString, parseISO } from 'utils/dates'
 import { RequestsContextProvider } from 'contexts/RequestsProvider'
 import { useBooleanState } from 'hooks/useBooleanState'
 import { LoadingModal } from 'components/LoadingModal'
+import { RouteProp, useFocusEffect, useNavigation, useRoute } from '@react-navigation/native'
+import { BottomTabNavigationType, BottomTabRoutes } from 'navigation/types'
+import { usePrevScreenBackHandler } from 'hooks/usePrevScreenBackHandler'
 import { CategoriesSlider } from './components/CategoriesSlider'
 
+type PrevScreen = 'NOTIFICATIONS' | undefined
```
Comment (iceu-bb) [inline src/screens/calendar/Calendar.tsx]: Please reuse `PrevScreens` type from `usePrevScreenBackHandler.ts` file

**Review by iceu-bb — CHANGES_REQUESTED** [review]

File: src/screens/calendar/Calendar.tsx (hunk 2 above)
Comment (iceu-bb) [reply src/screens/calendar/Calendar.tsx]: Thanks for the videos. It's a bad design if components as part of the internal state keep the previous screen name :D 
Please get it directly from navigation.

Comment (iceu-bb) [reply src/screens/calendar/Calendar.tsx]: You can check if the previous screen was "NOTIFICATIONS" inside `usePrevScreenBackHandler.ts` hook

**Review by iceu-bb — APPROVED** [review]

**Pull request merged**